docs/design/crd-proposal-phase1.md

```diff
@@ -0,0 +1,215 @@
+
+# Introduction
+
+This document is a proposal for changes and enhancements of the
+CRDs (Custom Resource Defintions) of the operator. I'm calling the
+current state of the CRDs "Phase 0" and the next set of CRDs "Phase 1".
+
+The current, Phase 0, CRDs support two features. One can define a
+SmbService CR that specifies a PVC name and it will create services
+to export one share, named "share", based on the PVC name. One can
+define an SmbPvc CR that embeds the PVC spec. This will automatically
+create a PVC and a matching SmbService CR to export it.
+
+
+# Phase 1 Custom Resources
+
+The following CRD types are proposed for Phase 1:
+* SmbShare - A CR that encapsulates all the information needed to export a single share
+* SmbSecurityDomain - A CR that encapsulates the knowledge needed to create "local" users or bind to an Active Directory domain
+
+The operator will take the SmbShare and SmbSecurityDomain resources as inputs
+and create as many smbd, winbind, etc. backing services as needed. Users
+will have limited input into what backing services the operator will create.
+The operator may, or may not, combine one or more share into a single smbd
+instance.
+
+SmbShare CRs will refer to one or more SmbSecurityDomain resources. The
+SmbSecurityDomain refernce will define the main security properties of the
+smbd that hosts the share. If two SmbShares are defined with different
+SmbSecurityDomains they may not be combined under one smbd.
+
+The listings below are not meant to be entirely complete but they outline
+the genearal direction to make the operaterator a fully-fledged tool to
+create shares that can be consumed by a variety of clients for various use
+cases. Some sections are deliberately marked "TBD" (to be decided) as
+placeholders for more detail that will be needed even in Phase 1.
+
+## SmbSecurityDomain
+
+Spec Options:
+* `mode` - enumerated string "user" or "active-directory" - Indicates that this is an AD joined instance or based on a local mapping of users
+* `users` - mapping - A new subsection pertaining to "user" options - required when mode = "user"
+  * `secret` - string - The name of a secret containing user and group definitions that will be used to create the local users and groups for the service
+* `localName` - string - Optional string used to indicate how this instance will
+  be referred to in an Active Directory
+* `realm` - string - The dns domain name of an Active Directory domain
+* `domains` - list-of-maps - A new subsection pertaining to detailed "active-directory" configuration (optional)
```

Comment:
It is supposed to contain as many settings for domains as you have (primary or trusted) that you want to configure.

Example:
```
domains:
  - domainName: myrealm.mycorp.example.com
    backend: ad
  - domainName: atrust.mycorp.example.com
    backend: ad
```

Do you you think it would be clearer as a map-of-maps like this example:
```
domains:
    myrealm.mycorp.example.com:
        backend: ad
    atrust.mycorp.example.com:
        backend: ad
```?

If we need to the section could include other parameters such as idmap ranges, etc. But I left it out of the initial draft becase (a) I was hoping to keep this siimple and (b) hoping the operator could do the heavy lifting of determining valid ranges after being simply told to use the ad (and check for rfc2307 in the ad).

Reply:
I think a tag for the name is fine, but `domainName` under `domains` is kind of redundant. we could just have `name` under `domains`?

Reply:
Sure. I thought a more descriptive key would be better but it is somewhat redundant.

Reply:
changed.